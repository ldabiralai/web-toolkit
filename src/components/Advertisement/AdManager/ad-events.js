export const REFRESH_AD_SLOT = 'WEBTOOLKIT_GPT_ADINIT_REFRESH_AD_SLOT';
export const INJECT_AD_SLOT = 'WEBTOOLKIT_GPT_ADINIT_INJECT_AD_SLOT';
