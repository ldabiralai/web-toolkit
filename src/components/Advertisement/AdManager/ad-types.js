export const BANNER = 'banner';
export const MPU = 'mpu';
export const BANNER_SPONSORSHIP = 'banner-sponsorship';
export const BETTING_BOX = 'betting-box';
export const INTERSTITIEL = 'interstitiel';
export const OUTSTREAM = 'outstream';
