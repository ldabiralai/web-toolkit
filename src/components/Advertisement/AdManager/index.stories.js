import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import readme from './readme.md';

storiesOf('Components|Advertisement', module)
  .addParameters({
    backgrounds: { disable: true },
  })
  .add('AdManager', withInfo({ text: readme })(() => <div>Ad manager Module - Click Show info</div>));
