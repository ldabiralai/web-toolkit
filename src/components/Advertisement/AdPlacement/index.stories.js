import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { select } from '@storybook/addon-knobs';
import { AdPlacement, AdManager } from '../../..';

const indexStories = storiesOf('Components|Advertisement', module).addParameters({
  backgrounds: { disable: true },
});

indexStories.add(
  'AdPlacement',
  withInfo()(() => (
    <>
      <AdPlacement adType={select('adType', Object.values(AdManager.enums.adTypes), AdManager.enums.adTypes.MPU)} />
    </>
  ))
);
