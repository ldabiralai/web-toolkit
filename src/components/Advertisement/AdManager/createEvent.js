import * as events from './ad-events';

export default (getGooglePublisherTagInstance, type, detail) => {
  let event;

  switch (type) {
    case events.INJECT_AD_SLOT:
      event = new CustomEvent(events.INJECT_AD_SLOT, { detail });
      break;
    case events.REFRESH_AD_SLOT:
      event = new CustomEvent(events.REFRESH_AD_SLOT, { detail });
      break;
    default:
      break;
  }

  if (!event) return;

  const fn = () => setTimeout(() => window.dispatchEvent(event), 0);

  if (getGooglePublisherTagInstance.apiReady) {
    fn();
  } else {
    // using GPT queue to start emiting window events
    getGooglePublisherTagInstance.cmd.push(fn);
  }
};
