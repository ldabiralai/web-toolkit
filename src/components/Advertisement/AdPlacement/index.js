import React from 'react';
import PropTypes from 'prop-types';
import AdManager from '../AdManager';

const AdPlacement = React.forwardRef(({ adType, isNoDesktop, isNoTablet, isNoMobile }, ref) => {
  const randomId = `adId${Date.now() + Math.floor(Math.random() * 1000)}`;

  return (
    <div data-adplacement-id={randomId} data-class="ad" ref={ref}>
      {AdManager.manageAds.injectAdSlot(adType, randomId, isNoMobile, isNoTablet, isNoDesktop)}
    </div>
  );
});

AdPlacement.defaultProps = {
  isNoTablet: false,
  isNoDesktop: false,
  isNoMobile: false,
};

AdPlacement.propTypes = {
  adType: PropTypes.oneOf([
    AdManager.enums.adTypes.MPU,
    AdManager.enums.adTypes.BETTING_BOX,
    AdManager.enums.adTypes.INTERSTITIEL,
    AdManager.enums.adTypes.OUTSTREAM,
    AdManager.enums.adTypes.BANNER_SPONSORSHIP,
    AdManager.enums.adTypes.BANNER,
  ]).isRequired,
  isNoTablet: PropTypes.bool,
  isNoDesktop: PropTypes.bool,
  isNoMobile: PropTypes.bool,
};

export default AdPlacement;
