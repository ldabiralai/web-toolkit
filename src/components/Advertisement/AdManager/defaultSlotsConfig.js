import * as adTypes from './ad-types';
import { RESULT, ARTICLE, LIVESTREAM, VIDEO, HP_SECTION, HP_MAIN, LIVE, OTHER } from './page-types';

const getSizes = (desktop, tablet, mobile) => ({
  desktop,
  'tablet-web': tablet,
  'mobile-web': mobile,
});

const multiple = (creative, times) => ({
  custom: creative,
  count: times,
});

const buildBannerObj = (...args) => {
  const banner = {};

  args.forEach(creativeObj => {
    let creative = creativeObj;

    let count = 1;

    if (creativeObj.custom) {
      creative = creativeObj.custom;
      // eslint-disable-next-line prefer-destructuring
      count = creativeObj.count;
    }

    if (banner[creative.name]) return;

    banner[creative.name] = {
      creative,
      count,
    };
  });

  return banner;
};

const creatives = {
  BANNER: {
    name: adTypes.BANNER,
    sizes: getSizes(null, null, [320, 50]),
  },

  BANNER_SPONSORSHIP: {
    name: adTypes.BANNER_SPONSORSHIP,
    sizes: getSizes([[970, 250], [970, 90], [728, 90]], [728, 90], [320, 50]),
  },

  MPU: {
    name: adTypes.MPU,
    sizes: getSizes([[300, 250], [300, 600], 'fluid'], [[300, 250], 'fluid'], [[300, 250], 'fluid']),
  },

  BETTING_BOX: {
    name: adTypes.BETTING_BOX,
    sizes: getSizes([[300, 170], 'fluid'], [[300, 170], 'fluid'], null),
  },

  INTERSTITIAL: {
    name: adTypes.INTERSTITIEL,
    sizes: getSizes(['out-of-page'], ['out-of-page'], ['out-of-page']),
  },

  OUTSTREAM: {
    name: adTypes.OUTSTREAM,
    sizes: getSizes([[1, 1], 'fluid'], [[1, 1], 'fluid'], [[1, 1], 'fluid']),
  },
};

const config = {
  desktop: {
    [HP_MAIN]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      multiple(creatives.MPU, 3),
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL
    ),
    [HP_SECTION]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      multiple(creatives.MPU, 3),
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL
    ),
    [VIDEO]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.BETTING_BOX, creatives.INTERSTITIAL),
    [LIVESTREAM]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.INTERSTITIAL),
    [ARTICLE]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      creatives.MPU,
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL,
      creatives.OUTSTREAM
    ),
    [LIVE]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      creatives.MPU,
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL,
      multiple(creatives.OUTSTREAM, 10)
    ),
    [RESULT]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      creatives.MPU,
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL
    ),
    [OTHER]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.BETTING_BOX, creatives.INTERSTITIAL),
  },
  'tablet-web': {
    [HP_MAIN]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      multiple(creatives.MPU, 3),
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL
    ),
    [HP_SECTION]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      multiple(creatives.MPU, 3),
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL
    ),
    [VIDEO]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.BETTING_BOX, creatives.INTERSTITIAL),
    [LIVESTREAM]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.INTERSTITIAL),
    [ARTICLE]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      creatives.MPU,
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL,
      creatives.OUTSTREAM
    ),
    [LIVE]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      multiple(creatives.MPU, 10),
      creatives.BETTING_BOX,
      creatives.INTERSTITIAL,
      multiple(creatives.OUTSTREAM, 10)
    ),
    [RESULT]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.INTERSTITIAL),
    [OTHER]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.INTERSTITIAL),
  },
  'mobile-web': {
    [HP_MAIN]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      multiple(creatives.BANNER, 3),
      multiple(creatives.MPU, 3),
      creatives.INTERSTITIAL
    ),
    [HP_SECTION]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      multiple(creatives.BANNER, 3),
      multiple(creatives.MPU, 3),
      creatives.INTERSTITIAL
    ),
    [VIDEO]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.INTERSTITIAL),
    [LIVESTREAM]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.INTERSTITIAL),
    [ARTICLE]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.MPU, creatives.INTERSTITIAL, creatives.OUTSTREAM),
    [LIVE]: buildBannerObj(
      creatives.BANNER_SPONSORSHIP,
      multiple(creatives.BANNER, 10),
      creatives.MPU,
      creatives.INTERSTITIAL,
      multiple(creatives.OUTSTREAM, 10)
    ),
    [RESULT]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.BANNER, creatives.INTERSTITIAL),
    [OTHER]: buildBannerObj(creatives.BANNER_SPONSORSHIP, creatives.BANNER, creatives.INTERSTITIAL),
  },
};

export default config;
