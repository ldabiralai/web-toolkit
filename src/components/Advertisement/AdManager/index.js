import customEventPolyfill from '../AdInit/utilities/customEventPolyfill';
import createEvent from './createEvent';
import defaultSlotsConfig from './defaultSlotsConfig';
import * as events from './ad-events';
import * as adTypes from './ad-types';
import * as pageTypes from './page-types';

customEventPolyfill();

const getGooglePublisherTagInstance = () => {
  window.googletag = window.googletag || {};
  window.googletag.cmd = window.googletag.cmd || [];

  return window.googletag;
};

const injectAdSlot = (adSlotType, placementId, isNoMobile, isNoTablet, isNoDesktop) => {
  createEvent(getGooglePublisherTagInstance(), events.INJECT_AD_SLOT, {
    adSlotType,
    placementId,
    isNoMobile,
    isNoTablet,
    isNoDesktop,
  });
};

const refreshAdSlot = slotId => {
  createEvent(getGooglePublisherTagInstance(), events.REFRESH_AD_SLOT, { slotId });
};

const registerInjectAdEvent = cb => {
  window.addEventListener(events.INJECT_AD_SLOT, cb);
};

const registerRefreshAdEvent = cb => {
  window.addEventListener(events.REFRESH_AD_SLOT, cb);
};

export default {
  enums: {
    adTypes,
    pageTypes,
    events,
  },
  registerEvents: {
    injectAd: registerInjectAdEvent,
    refreshAd: registerRefreshAdEvent,
  },
  manageAds: {
    injectAdSlot,
    refreshAdSlot,
  },

  googlePublisherTag: {
    getInstance: getGooglePublisherTagInstance,
  },
  defaultSlotsConfig,
};
