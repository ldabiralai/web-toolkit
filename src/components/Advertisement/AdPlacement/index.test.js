import React from 'react';
import { shallow } from 'enzyme';
import AdPlacement from '.';
import AdManager from '../AdManager';

jest.mock('../AdManager');

describe('AdPlacement', () => {
  let globDate;
  let globMath;

  beforeAll(() => {
    globDate = global.Date;
    globMath = global.Math;

    delete global.Date;

    global.Date = {
      now: () => 'now',
    };

    const mockMath = Object.create(global.Math);
    mockMath.random = () => 0.5;
    global.Math = mockMath;
  });

  afterAll(() => {
    global.Date = globDate;
    global.Math = globMath;
    jest().resetAllMocks();
  });

  it('renders ad with random div', () => {
    expect(shallow(<AdPlacement adType="mpu" isNoDesktop isNoMobile isNoTablet />)).toMatchSnapshot();
  });

  it('calls adManager to render the given ad', () => {
    AdManager.manageAds.injectAdSlot.mockResolvedValue(jest.fn());
    shallow(<AdPlacement adType="betting-box" isNoDesktop isNoMobile isNoTablet />);
    expect(AdManager.manageAds.injectAdSlot).toHaveBeenCalledWith('betting-box', 'adIdnow500', true, true, true);
  });
});
