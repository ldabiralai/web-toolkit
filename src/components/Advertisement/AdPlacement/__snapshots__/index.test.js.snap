// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdPlacement renders ad with random div 1`] = `
<div
  data-adplacement-id="adIdnow500"
  data-class="ad"
/>
`;
